# WE04 DecisonTrees
# Bapuji Satyala 
# U61600601

In this Notebook we have cosidered a dataset from the UCI repository, where this dataset has 2139 observations with 24 features and 1 target, where our target variable is Binary column. This is a Classification Problem where  we are predicting if the patient have died within a certain window of time or not with diagnosed as AIDS

The link for the dataset : https://archive.ics.uci.edu/dataset/890/aids+clinical+trials+group+study+175

## Importing the required libraries that we will be using in this note book.

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score,f1_score, precision_score, recall_score
from ucimlrepo import fetch_ucirepo
from sklearn.preprocessing import FunctionTransformer
from sklearn.model_selection import GridSearchCV
np.random.seed(1)

To import the data we need to install the ucimlrepo as the dataset is Present in that repo, commented the below statement as the repo is already installed.

In [2]:
#pip install ucimlrepo

Fetching the dataset from the repo and the id number of the dataset is 890 and we are saving the data set to the dataframe df 

In [3]:
AD = fetch_ucirepo(id=890) 
df = AD.data.original 


Now we are looking at the top 5 observations of the dataframe which contain all the features and the target

In [4]:
df.head()

pidnum  cid  time  trt  age     wtkg  hemo  homo  drugs  karnof  ...  \
0   10056    0   948    2   48  89.8128     0     0      0     100  ...   
1   10059    1  1002    3   61  49.4424     0     0      0      90  ...   
2   10089    0   961    3   45  88.4520     0     1      1      90  ...   
3   10093    0  1166    3   47  85.2768     0     1      0     100  ...   
4   10124    0  1090    0   43  66.6792     0     1      0     100  ...   

   gender  str2  strat  symptom  treat  offtrt  cd40  cd420  cd80  cd820  
0       0     0      1        0      1       0   422    477   566    324  
1       0     1      3        0      1       0   162    218   392    564  
2       1     1      3        0      1       1   326    274  2063   1893  
3       1     1      3        0      1       0   287    394  1590    966  
4       1     1      3        0      0       0   504    353   870    782  

[5 rows x 25 columns]

We are looking at the number of rows and columns in the data frame.The dataframe contains 2139 observations and 25 columns 

In [5]:
df.shape

(2139, 25)

Checking if there are any null values in the data frame and additing them to get more clear understanding on data if there are any nulls in the data frame or not

In [6]:
df.isna().sum()

pidnum     0
cid        0
time       0
trt        0
age        0
wtkg       0
hemo       0
homo       0
drugs      0
karnof     0
oprior     0
z30        0
zprior     0
preanti    0
race       0
gender     0
str2       0
strat      0
symptom    0
treat      0
offtrt     0
cd40       0
cd420      0
cd80       0
cd820      0
dtype: int64

Lets look at the details of the data frame and their data types and the counts of non null values

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2139 entries, 0 to 2138
Data columns (total 25 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   pidnum   2139 non-null   int64  
 1   cid      2139 non-null   int64  
 2   time     2139 non-null   int64  
 3   trt      2139 non-null   int64  
 4   age      2139 non-null   int64  
 5   wtkg     2139 non-null   float64
 6   hemo     2139 non-null   int64  
 7   homo     2139 non-null   int64  
 8   drugs    2139 non-null   int64  
 9   karnof   2139 non-null   int64  
 10  oprior   2139 non-null   int64  
 11  z30      2139 non-null   int64  
 12  zprior   2139 non-null   int64  
 13  preanti  2139 non-null   int64  
 14  race     2139 non-null   int64  
 15  gender   2139 non-null   int64  
 16  str2     2139 non-null   int64  
 17  strat    2139 non-null   int64  
 18  symptom  2139 non-null   int64  
 19  treat    2139 non-null   int64  
 20  offtrt   2139 non-null   int64  
 21  cd40     2139 

We are looking at the statstics of the data in the dataframe, from the below we can see the minimum values, maximum values and the mean and standard deviadtion in each column of the data frame 

In [8]:
df.describe()

pidnum          cid         time          trt          age  \
count    2139.000000  2139.000000  2139.000000  2139.000000  2139.000000   
mean   248778.252454     0.243572   879.098177     1.520804    35.248247   
std    234237.289399     0.429338   292.274324     1.127890     8.709026   
min     10056.000000     0.000000    14.000000     0.000000    12.000000   
25%     81446.500000     0.000000   727.000000     1.000000    29.000000   
50%    190566.000000     0.000000   997.000000     2.000000    34.000000   
75%    280277.000000     0.000000  1091.000000     3.000000    40.000000   
max    990077.000000     1.000000  1231.000000     3.000000    70.000000   

              wtkg         hemo         homo        drugs       karnof  ...  \
count  2139.000000  2139.000000  2139.000000  2139.000000  2139.000000  ...   
mean     75.125311     0.084151     0.661057     0.131370    95.446470  ...   
std      13.263164     0.277680     0.473461     0.337883     5.900985  ...   
min      31.000000     0.000000     0.000000     0.000000    70.000000  ...   
25%      66.679200     0.000000     0.000000     0.000000    90.000000  ...   
50%      74.390400     0.000000     1.000000     0.000000   100.000000  ...   
75%      82.555200     0.000000     1.000000     0.000000   100.000000  ...   
max     159.939360     1.000000     1.000000     1.000000   100.000000  ...   

            gender         str2        strat      symptom        treat  \
count  2139.000000  2139.000000  2139.000000  2139.000000  2139.000000   
mean      0.827957     0.585788     1.979897     0.172978     0.751286   
std       0.377506     0.492701     0.899053     0.378317     0.432369   
min       0.000000     0.000000     1.000000     0.000000     0.000000   
25%       1.000000     0.000000     1.000000     0.000000     1.000000   
50%       1.000000     1.000000     2.000000     0.000000     1.000000   
75%       1.000000     1.000000     3.000000     0.000000     1.000000   
max       1.000000     1.000000     3.000000     1.000000     1.000000   

            offtrt         cd40        cd420         cd80        cd820  
count  2139.000000  2139.000000  2139.000000  2139.000000  2139.000000  
mean      0.362786   350.501169   371.307153   986.627396   935.369799  
std       0.480916   118.573863   144.634909   480.197750   444.976051  
min       0.000000     0.000000    49.000000    40.000000   124.000000  
25%       0.000000   263.500000   269.000000   654.000000   631.500000  
50%       0.000000   340.000000   353.000000   893.000000   865.000000  
75%       1.000000   423.000000   460.000000  1207.000000  1146.500000  
max       1.000000  1199.000000  1119.000000  5011.000000  6035.000000  

[8 rows x 25 columns]

### Step 2: Dropping the  variables that are not relavant to our analysis 

The race column do hold data for the patients if they are White or Non-White which is not relavant to our analysis, so we are dropping the column from our data frame.

In [9]:
df.drop('race', axis=1, inplace=True)

Lets look at the number of columns after dropping the race column from the dataframe

In [10]:
df.shape

(2139, 24)

### Step 3: Splitting the data into Train set and the Test set  where our target column is still not seperated.

We are splitting the dataset into test and train sets by splitting the data into 70 % of Train data and 30% as the test data

In [11]:
train_set, test_set = train_test_split(df, test_size=0.3)

### Step 4: Checking for the Missing Values in the dataset

In [12]:
train_set.isna().sum()

pidnum     0
cid        0
time       0
trt        0
age        0
wtkg       0
hemo       0
homo       0
drugs      0
karnof     0
oprior     0
z30        0
zprior     0
preanti    0
gender     0
str2       0
strat      0
symptom    0
treat      0
offtrt     0
cd40       0
cd420      0
cd80       0
cd820      0
dtype: int64

Here we can see that there are no missing values in the train dataset which has 70% of the data

In [13]:
test_set.isna().sum()

pidnum     0
cid        0
time       0
trt        0
age        0
wtkg       0
hemo       0
homo       0
drugs      0
karnof     0
oprior     0
z30        0
zprior     0
preanti    0
gender     0
str2       0
strat      0
symptom    0
treat      0
offtrt     0
cd40       0
cd420      0
cd80       0
cd820      0
dtype: int64

Here we can see that there are no missing values in the test dataset which has 30% of the data

Seperating the target variable as we are transforming the datasets and we dont want to transform the target which is cid and we are seperating the each sets(Train,Test) into the train_y,test_y and train_inputs,test_inputs

In [14]:
train_y = train_set['cid']
test_y = test_set['cid']

train_inputs = train_set.drop(['cid'], axis=1)
test_inputs = test_set.drop(['cid'], axis=1)


### Step 5: Feature Engineering a new column

Here we are deriving a new column called the Elder based on the existing column age where we are considering it as 0 if the age is greater than or equal to 65 if the age is lesser than 65 we are considering it as 1. We are duplicating the dataframe to maintain the original structure as we are introducing a new column.

In [15]:
def new_col(df):
    
    df1 = df.copy()
    
    df1['Elder'] = df1['age'].apply(lambda x: '0' if x >= 65 else '1' )
    
    return df1[['Elder']]


We are calling the new_col function we have created on the train_set

In [16]:
new_col(train_set)


Elder
19       1
622      1
1975     1
250      1
2042     1
...    ...
960      1
905      1
1096     1
235      1
1061     1

[1497 rows x 1 columns]

### Step 6: Identifying the Types of columns 

We are displaying the type of train_inputs, here we can see that all the the variables are of numeric type.

In [17]:
train_inputs.dtypes

pidnum       int64
time         int64
trt          int64
age          int64
wtkg       float64
hemo         int64
homo         int64
drugs        int64
karnof       int64
oprior       int64
z30          int64
zprior       int64
preanti      int64
gender       int64
str2         int64
strat        int64
symptom      int64
treat        int64
offtrt       int64
cd40         int64
cd420        int64
cd80         int64
cd820        int64
dtype: object

We are identifying the numeric columns in the train_input set by using the select_dtypes method

In [18]:
numeric_columns = train_inputs.select_dtypes(include=[np.number]).columns.to_list()

Identifying the binary columns

In [19]:
binary_columns = ['hemo', 'homo','drugs','oprior','z30','zprior','gender','str2','symptom','treat','offtrt']

We are identifying the binary columns in the numeric columns and removing them from the numeric columns as the binary columns will be read as the numerics, hence we are removing them from the numeric columns 

In [20]:
for col in binary_columns:
    numeric_columns.remove(col)

Displaying the numeric Columns after removing the binary columns 

In [21]:
numeric_columns

['pidnum',
 'time',
 'trt',
 'age',
 'wtkg',
 'karnof',
 'preanti',
 'strat',
 'cd40',
 'cd420',
 'cd80',
 'cd820']

Displaying the binary columns 

In [22]:
binary_columns 

['hemo',
 'homo',
 'drugs',
 'oprior',
 'z30',
 'zprior',
 'gender',
 'str2',
 'symptom',
 'treat',
 'offtrt']

In [23]:
new_cols = ['age']

### Step 7: Creating Transformer Pipelines for each type columns 


The Pipelines are used to flexibilly chain the preprocessing steps such as the scaling, encoding and also the model evaluations can also be done, this benefits in preventing from data leakage  and the code reusability can also be done as these objects(estimator objects) are modular.The Transformers implement the fitting and transorm methods. In The below we are using the Standard scaler to scale the values. Here we are creating pipeline for the binary transformer where we do have single step for standardizing the binary columns where we will be further using this pipeline to the data and scale the binary columns, we have not cosidered the categorical transformer as we dont have any categorical columns in our dataset.

In [24]:
numeric_transformer = Pipeline(steps=[
                ('scaler', StandardScaler())])

In [25]:
binary_transformer = Pipeline(steps=[('scaler', StandardScaler())])

We are ingesting the new column which we have created during feature engineering the functional transformer does this custom function and we are also scaling the column by using the standardscaler

In [26]:
my_new_column = Pipeline(steps=[('my_new_column', FunctionTransformer(new_col)),
                                 ('scaler', StandardScaler())])

### Step 8: Combining all the Transformer Pipelines we have created

The ColumnTransformer will be used when the dataset consists of differnt types of data types.we will apply all the transformations to different columns in the dataset,in the below code we are applying binary transformation, numeric transformation to the data set.The remainder is set to 'passthrough' which means if there are any columns that were not transformed will be included in the output by not performing any transformation.

In [27]:
preprocessor = ColumnTransformer([
        ('num', numeric_transformer, numeric_columns),
        ('binary', binary_transformer, binary_columns),
        ('trans', my_new_column, new_cols)],
        remainder='passthrough')
preprocessor

ColumnTransformer(remainder='passthrough',
                  transformers=[('num',
                                 Pipeline(steps=[('scaler', StandardScaler())]),
                                 ['pidnum', 'time', 'trt', 'age', 'wtkg',
                                  'karnof', 'preanti', 'strat', 'cd40', 'cd420',
                                  'cd80', 'cd820']),
                                ('binary',
                                 Pipeline(steps=[('scaler', StandardScaler())]),
                                 ['hemo', 'homo', 'drugs', 'oprior', 'z30',
                                  'zprior', 'gender', 'str2', 'symptom',
                                  'treat', 'offtrt']),
                                ('trans',
                                 Pipeline(steps=[('my_new_column',
                                                  FunctionTransformer(func=<function new_col at 0x0000025389359260>)),
                                                 ('scaler', StandardScaler())]),
                                 ['age'])])

### Step 9:Fitting and transforming the combined Pipeline for the Train data 



The Preprocessor is a object of the column transformer which is assigned in the above step where we are fitting and transforming the training input data. Before applying any transformations the transformer needs to learn from the training data.once the transformer fitted the data then the transform will be used to transform the data.Here we are doing the fit and transform in single operation.

In [28]:
train_x = preprocessor.fit_transform(train_inputs)


Lets look at the shape of the train_x after ingesting the new column as we have seen before the features were 23(after removing the irrevelant column for the analysis) and after feature engineering we have included the new column so the shape has been changed to 24

In [29]:
train_x.shape

(1497, 24)

### Step 10: Fitting the combined Pipeline for the Test data 



Here we are transforming the test_inputs

In [30]:
test_x = preprocessor.transform(test_inputs)

test_x

array([[-2.94819806e-01,  5.76502227e-01, -4.46722513e-01, ...,
         5.84292606e-01, -7.41410440e-01,  6.34361480e-02],
       [-3.35380452e-01,  2.12507863e-01, -1.33013546e+00, ...,
        -1.71147126e+00,  1.34878058e+00,  6.34361480e-02],
       [-9.66446017e-01, -2.80490336e+00, -1.33013546e+00, ...,
        -1.71147126e+00, -7.41410440e-01,  6.34361480e-02],
       ...,
       [-2.75713271e-03,  8.62254812e-01, -4.46722513e-01, ...,
         5.84292606e-01, -7.41410440e-01,  6.34361480e-02],
       [ 8.41645332e-02,  7.29583969e-01, -1.33013546e+00, ...,
        -1.71147126e+00, -7.41410440e-01,  6.34361480e-02],
       [-8.44324402e-02,  7.56798501e-01,  4.36690435e-01, ...,
         5.84292606e-01, -7.41410440e-01,  6.34361480e-02]])

Lets look at the shape of the test_x after ingesting the new column as we have seen before the features were 23 and after feature engineering we have included the new column so the shape has been changed to 24

In [31]:
test_x.shape

(642, 24)

### Step 11: Creating a baseline using the dummy classifier

The Dummy Classifier is a simple classifier which serves as baseline model to compare the performance of complex models.We are setting a reference point against the other classification models.Here we are using the most_frequent as strategy whcih means it always predict the most frequent class.We are fitting the train data into the model and calculating the accuracy of the model

In [32]:
dummy_clf = DummyClassifier(strategy="most_frequent",random_state=1)

dummy_clf.fit(train_x, train_y)

DummyClassifier(random_state=1, strategy='most_frequent')

Here we are predicting the test data as the DummyClassifier is already trained with most_frequent strategy, it will always predicts the most frequent class observed during the training 

In [33]:
dummy_test_pred = dummy_clf.predict(test_x)
baseline_test_acc = accuracy_score(test_y, dummy_test_pred)
print('The Baseline Test Accuracy: {}' .format(baseline_test_acc))



The Baseline Test Accuracy: 0.7570093457943925


The Baseline test accuracy is 0.7570093457943925

### DecisionTree Classifier 

The Decision Trees are one of the most popularly used algorithms due to their simplicity and effectiveness whcih is a supervised learning algorithm Predictions are made by traversing the tree from root to leaf nodes based on feature values.The DT can be used for both regressor and classification problem as our problem is Classification we are using the DecisionTreeClassifier

Here we are specifying the max depth of the decision tree, this factor will control the maximum number of levels in the tree, which will prevent from over fitting. Here we are fitting the decision tree classifier to the train data.

In [34]:
tree_clf = DecisionTreeClassifier(max_depth=5,random_state=1)

tree_clf.fit(train_x, train_y)

DecisionTreeClassifier(max_depth=5, random_state=1)

Here the Model is predicting the train data there by using the trained decison tree classifier

In [35]:
train_y_pred = tree_clf.predict(train_x)
AC_score=accuracy_score(train_y, train_y_pred)
F_score=f1_score(train_y, train_y_pred)
r_score =recall_score(train_y, train_y_pred)
p_score= precision_score(train_y, train_y_pred)
print("The Accuracy :", AC_score)
print("The F1 Score :", F_score)
print("The Recall Score :", r_score)
print("The Precision Score :", p_score)

The Accuracy : 0.9205076820307281
The F1 Score : 0.8282828282828283
The Recall Score : 0.7863013698630137
The Precision Score : 0.875


Here the Model is predicting the unseen test data there by using the trained decison tree classifier

In [36]:

test_y_pred = tree_clf.predict(test_x)

AC_score=accuracy_score(test_y, test_y_pred)
F_score=f1_score(test_y, test_y_pred)
r_score =recall_score(test_y, test_y_pred)
p_score= precision_score(test_y, test_y_pred)
print("The Accuracy :", AC_score)
print("The F1 Score :", F_score)
print("The Recall Score :", r_score)
print("The Precision Score :", p_score)

The Accuracy : 0.883177570093458
The F1 Score : 0.7457627118644068
The Recall Score : 0.7051282051282052
The Precision Score : 0.7913669064748201


## Performing Regularization

Regularization is the technique to prevent the overfitting, here we have considered the min_samples_leaf as 10, setting a higher will result in few leaf nodes which will reduce the complexity of the tree and overfiting can be controled as these leaf nodes have more train instances associated with them. 

In [37]:
tree_clf2 = DecisionTreeClassifier(min_samples_leaf = 10, max_depth=5,random_state=1)

tree_clf2.fit(train_x, train_y)

DecisionTreeClassifier(max_depth=5, min_samples_leaf=10, random_state=1)

We are evaluating the Performance of the decision tree Classifier on the training data. and Priniting the Outputs

In [38]:
train_y_pred = tree_clf2.predict(train_x)

AC_score=accuracy_score(train_y, train_y_pred)
F_score=f1_score(train_y, train_y_pred)
r_score =recall_score(train_y, train_y_pred)
p_score= precision_score(train_y, train_y_pred)
print("The Accuracy :", AC_score)
print("The F1 Score :", F_score)
print("The Recall Score :", r_score)
print("The Precision Score :", p_score)

The Accuracy : 0.9084836339345357
The F1 Score : 0.8005822416302766
The Recall Score : 0.7534246575342466
The Precision Score : 0.8540372670807453


We are evaluating the Performance of the decision tree Classifier on the test data. and Priniting the Outputs, to check how well the model predicts onn the unseen data.

In [39]:

test_y_pred = tree_clf2.predict(test_x)

AC_score=accuracy_score(test_y, test_y_pred)
F_score=f1_score(test_y, test_y_pred)
r_score =recall_score(test_y, test_y_pred)
p_score= precision_score(test_y, test_y_pred)
print("The Accuracy :", AC_score)
print("The F1 Score :", F_score)
print("The Recall Score :", r_score)
print("The Precision Score :", p_score)

The Accuracy : 0.8738317757009346
The F1 Score : 0.7216494845360825
The Recall Score : 0.6730769230769231
The Precision Score : 0.7777777777777778


### Step 12: Implementing Grid searchCV  and testing the parameter combinations and fitting Decision Tree Classifier 

The Hyperparameter tuning is a necessary step for optimal model performance although it is a iterative and resource-intensive process.The performance of the model can be greatly impacted by tuning these hyperparameters by determining their ideal set of values. Further in the below i have considered the GridsearchCV where the gridsearch involves will be searching for the best parameters through a predefined set of them to get the best parameters for the best effeciency of the model.The gridsearch will evaluate exhaustively all possible combinations off the hyperparameters.

In [40]:
param_grid = {
              'max_depth' :  range(1,16),
              'criterion' :['gini', 'entropy'],
              'min_samples_leaf' : range(1,16)}             
tree_class = DecisionTreeClassifier(random_state=1)
grid_search = GridSearchCV(estimator=tree_class, param_grid=param_grid, cv=5, verbose=True,scoring='accuracy')
grid_search.fit(train_x, train_y)

Fitting 5 folds for each of 450 candidates, totalling 2250 fits


GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=1),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': range(1, 16),
                         'min_samples_leaf': range(1, 16)},
             scoring='accuracy', verbose=True)

Here we are printing the best parameters that the grid search have evaluated that gives the best performace of the DecisionTreeClassifier

In [41]:
print(grid_search.best_params_)

{'criterion': 'entropy', 'max_depth': 4, 'min_samples_leaf': 4}


We are evaluating the Performance of the decision tree Classifier on the test data. and Priniting the Outputs, to check how well the model predicts on the unseen data.

In [42]:
pred_y = grid_search.predict(test_x)
print("   DecisionTree GridSearch Metric Scores   ")
accuracy = accuracy_score(test_y, pred_y)
print("The Accuracy is  :", accuracy)
f1 = f1_score(test_y, pred_y)
print("The F1 Score is :", f1)
recall = recall_score(test_y, pred_y)
print("The Recall is :", recall)
precision = precision_score(test_y, pred_y)
print("The Precision is :", precision)


   DecisionTree GridSearch Metric Scores   
The Accuracy is  : 0.8582554517133957
The F1 Score is : 0.6872852233676975
The Recall is : 0.6410256410256411
The Precision is : 0.7407407407407407


### Summary

I have Considered a dataset from the UCI repository which holds the data of the patients that are diagnosed with AIDS, by comnsidering the two types of treatments that patients have recieved and the task is to predict whether each patient died within certain window or not.The dataset consisted of 24 features and one target where our target is the censory indicator which inidicates that the patient have died or not, it is a classification problem, i have dropped the variables that are not relavant to the analysis and splitted the data into 70% of the train set and the 30% as the test set and the preporocessing of the data there by feature engineered a new column 'Elder' based on the existing column there by considering condition if the age of patient is greater or lesser than 65.Based on the type of the data the pipelines are built to perfrom specified steps such as scaling and used the function transformer to ingest the newly feature engineered column and by scaling the values within it.A baseline model (dummy classfier) was fit to compare and evaluate the Performance of the Decisiontree classifer by using the Gridsearch hyperparameter tuning. The baseline has been set to 0.757 as the Dummy classifer achieved accuracy as the 0.757.The scores that are mentioned in the below are the scores of the models that have predicted the test data which is unseen by the model.

The Metric scores of the models are as follows:

### Decision Tree Classification Model

- The Accuracy  : 0.883177570093458
- The F1 Score  : 0.7457627118644068
- The Recall Score : 0.7051282051282052
- The Precision Score : 0.7913669064748201

The Regularization has been done to the Decision Tree Classifier where we have introduced the min_samples_leaf = 10, max_depth=5, the regularization is performed these regularization parameters will prevent it from overfitting.The Metrice scores after regularization are as below.

- The Accuracy : 0.8738317757009346
- The F1 Score : 0.7216494845360825
- The Recall Score : 0.6730769230769231
- The Precision Score : 0.7777777777777778

The Hyperparameter Tuning has been performed to check and test the different set of parameters we have defined where i have Considered the maximum depth will range from 1 to 16 and the criterion will been considered  between the gini and the entropy  where both are used to evaluate the purity of a node.The DT algorithm uses these criterions inorder to determine the best split at each node in the process of growing the tree and the minimum samples leaf will be checked between 1 and 16 which specifies the minimumm number of samples that are required to be present in leaf node.
The below are the best set of parameters and the metric scores that are evaluated from the GRid search of the decision tree Classifier

{'criterion': 'entropy', 'max_depth': 4, 'min_samples_leaf': 4}

- The Accuracy is  : 0.8582554517133957
- The F1 Score is : 0.6872852233676975
- The Recall is : 0.6410256410256411
- The Precision is : 0.7407407407407407


However,even The Accuracy of the DecisionTree classifier with only max_depth=5 hold the highest accuracy, there might be a sign of overfitting, once the reguularization has been done to the model by ingesting the min_samples_leaf = 10 the accuracy score has been reduced and eventually all the metrices like the F1score, Precision and the recall has been reduced which clearly shows that the normal Decision Tree classifer with only the max depth is 5 is overfitting.After Hyper parameter tuning the grid search has evaluated the best criterion as entropy and the Accuracy score has been reduced from 0.87 to 0.85 and the  F1 score has been reduced from 0.72 to 0.68, the scores of recall and precision have also reduced which can be sign of overfitting. 

When compared with the baseline model which is the simplifed model which makes predictions with the stratified as the strategy which predicts the class labels based on their distributions holds the accuracy of 0.75 which is quite lower when compared to that of the DecisionTree classifier with the Gridsearch(Hyper parameter tuning), the Gridsearched DecisionTree Classifier outperformed the Dummy classifier perfromance, we also know that the baseline is used to compare the complicated models,if the complicated models cannot outperform the baseline that means that the model is not learning from the data.In our case all the models outperfromed the baseline model and from them the Gridsearched Decision Tree classifier held as the best model when compared the metrices and by considering diferent parameters.